include/mata/inter-aut.hh

```diff
@@ -0,0 +1,193 @@
+ */
+struct FormulaNode
+{
+public:
+    enum OperandType {
+        SYM,
+        STATE,
+        NODE,
+        NOT_OPERAND
+    };
+
+    enum OperatorType {
+        NEG,
+        AND,
+        OR,
+        NOT_OPERATOR
+    };
+
+    enum Type {
+        OPERAND,
+        OPERATOR,
+        LEFT_PARENTHESIS,
+        RIGHT_PARENTHESIS,
+        UNKNOWN
+    };
+
+    /// Define whether a node is operand or operator
+    Type type;
+    /// Raw name of node as it was specified in input text, i.e., including type marker.
+    std::string raw;
+    /// Parsed name, i.e., a potential type marker (first character) is removed.
+    std::string name; // parsed name. When type marking is used, markers are removed.
+    /// if a node is operator, it defines which one
+    OperatorType operator_type;
+    /// if a node is operand, it defines which one
+    OperandType operand_type;
+
+    bool is_operand() const { return type == Type::OPERAND; }
+
+    bool is_operator() const { return type == Type::OPERATOR; }
+
+    bool is_righpar() const { return type == Type::RIGHT_PARENTHESIS; }
+
+    bool is_leftpar() const { return type == Type::LEFT_PARENTHESIS; }
+
+    FormulaNode() : type(UNKNOWN), raw(""), name(""), operator_type(NOT_OPERATOR), operand_type(NOT_OPERAND) {}
+
+    FormulaNode(Type t, std::string raw, std::string name,
+                OperatorType oprtor) : type(t), raw(raw), name(name), operator_type(oprtor),
+                                       operand_type(NOT_OPERAND) {}
+
+    FormulaNode(Type t, std::string raw, std::string name,
+                OperandType operand) : type(t), raw(raw), name(name), operator_type(NOT_OPERATOR),
+                                       operand_type(operand) {};
+
+    FormulaNode(Type t, std::string raw) : type(t), raw(raw), name(raw), operator_type(NOT_OPERATOR),
+                                           operand_type(NOT_OPERAND) {};
+};
+
+/**
+ * Structure representing a transition formula using a graph.
+ * A node of graph consists of node itself and set of children nodes.
+ * Nodes are operators and operands of the formula.
+ * E.g., a formula q1 & s1 will be transformed to a tree with & as a root node
+ * and q1 and s2 being children nodes of the root.
+ */
+struct FormulaGraph
+{
+    FormulaNode node;
+    std::vector<FormulaGraph> children;
+
+    FormulaGraph() : node(), children() {}
+    FormulaGraph(FormulaNode n) : node(n), children() {}
+
+    std::unordered_set<std::string> collect_node_names() const;
+};
+
+/**
+ * Structure for a general intermediate representation of parsed automaton.
+ * It contains information about type of automata, type of naming of nodes, symbols and states
+ * and type of alphabet. It contains also the transitions formula and formula for initial and final
+ * states. The formulas are represented as tree where nodes are either operands or operators.
+ */
+struct InterAutomaton
```

Comment:
I'm not really fond of this name. `Inter` could refer to several meanings, I would prefer something different. Mabye AutomatonRepresentation? ParsedAutomaton?

Reply:
I have been wondering about this, too. I would not guess `intermediate` from the shortened name. Could we maybe use `IntermediateAut`, for example?

Reply:
It is tooooo loooooong. You write it often, please something shorter guys.

Reply:
Well, I use autocompleters, so I have no issue with long names ;)

Reply:
Then I agree with the current `InterAutomaton`. Why not just `Aut`, btw.? If we want to spare our users from some additional typing… As of now, I do not have a better idea for the name.

Reply:
Or just `IA` (trollface)

Reply:
I think, we should rename all variables, classes and structure to `a`, `b`, ..., `z`, and then `aa`, `ab`....

Reply:
so this could be `a`

Reply:
renamed to `IntermediateAut`. It is only one character longer, I can live with it.